### PR TITLE
Support for any scalar coefficient in partial assembly operators [drzisga/pa_coeff]

### PR DESCRIPTION
**Pull request by drzisga**
This PR adds support for any Coefficient class implementation in the partial assembly of the mass and diffusion operators. The test checks that the result of a matrix-vector multiplication with the partially assembled operator is the same as using the assembled matrix.

The OCCA kernels are updated too, but I could not test them myself yet.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/1107 | @drzisga  | @tzanio | @dmed256 + @artv3 | 11/23/19 | 12/5/19 | ⌛due 12/12/19 |

### CODE REVIEW

**Comment by camierjs** [comment]
Even if we construct a `Vector coeff` of size `nq*ne` in the case of a scalar coefficient, it makes the API cleaner. I'm not sure it is worth it to bundle a `bool` and a `double` down to all the different implementations of the setup kernels.
@v-dobrev, do you have an elegant solution for this?

**Comment by camierjs** [comment]
I'm fine with this solution, that's a good way for these setup kernels!

**Comment by drzisga** [comment]
I implemented it like discussed, but unfortunately, the OCCA interface had to be changed by adding a boolean variable. I also think it would be possible to remove the conditional branch in the OCCA kernels by adding some preprocessor flag using `occa::properties` and adding some `#ifdefs` in the kernel. This, however, will make the kernels way more complicated and two different kernels will be compiled for the constant and non-constant coefficient case.

Again, I could not test the modified OCCA kernels.

File: tests/unit/fem/test_pa_coeff.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
```
Comment (tzanio) [inline tests/unit/fem/test_pa_coeff.cpp]: Thanks for adding this test!

**Comment by tzanio** [comment]
ping: @artv3

**Comment by artv3** [comment]
This looks good to me. I haven't gotten a chance to do try additional user scaling (https://github.com/mfem/mfem/pull/1141) , but I suspect it might be a minor extension if needed. I can make a PR if anything is missing.

**Comment by tzanio** [comment]
Very useful extension, thanks @drzisga!

**Pull request merged**